src/Pact/Analyze/Types.hs

```diff
@@ -912,86 +929,42 @@ userShow = userShowsPrec 0
 -- useful.
-data Invariant a where
-
-  -- literals
-  -- | A 'Decimal' literal
-  IDecimalLiteral :: Decimal -> Invariant Decimal
-  -- | An 'Integer' literal
-  IIntLiteral     :: Integer -> Invariant Integer
-  -- | A 'String' literal
-  IStringLiteral  :: Text    -> Invariant String
-  -- | A 'Time' literal
-  ITimeLiteral    :: Time    -> Invariant Time
-  -- | A 'Bool' literal
-  IBoolLiteral    :: Bool    -> Invariant Bool
-
-  -- | Creates an invariant expression from a symbolic value
-  ISym :: S a -> Invariant a
-
-  -- variables
+data Invariant a
+  = PureInvariant (PureTerm EInvariant Invariant a)
+  -- TODO: we can move this to PureTerm when we assign invariant vars `VarId`s
   -- | Refers to a variable with the provided name
-  IVar :: Text -> Invariant a
+  | IVar Text
+  deriving (Show, Eq)
 
-  -- string ops
-  -- | The concatenation of two 'String' expressions
-  IStrConcat :: Invariant String -> Invariant String -> Invariant String
-  -- | The length of a 'String' expression
-  IStrLength :: Invariant String                     -> Invariant Integer
+instance S :<: Invariant where
+  inject = PureInvariant . Sym
+  project = \case
+    PureInvariant (Sym a) -> Just a
+    _                     -> Nothing
 
-  -- numeric ops
-  -- | An 'ArithOp' expression over two 'Decimal' expressions
-  IDecArithOp      :: ArithOp             -> Invariant Decimal -> Invariant Decimal -> Invariant Decimal
-  -- | An 'ArithOp' expression over two 'Integer' expressions
-  IIntArithOp      :: ArithOp             -> Invariant Integer -> Invariant Integer -> Invariant Integer
-  -- | A 'UnaryArithOp' expression over a 'Decimal' expression
-  IDecUnaryArithOp :: UnaryArithOp        -> Invariant Decimal ->                      Invariant Decimal
-  -- | A 'UnaryArithOp' expression over an 'Integer expression
-  IIntUnaryArithOp :: UnaryArithOp        -> Invariant Integer ->                      Invariant Integer
-  -- | An 'ArithOp' expression over a 'Decimal' and an 'Integer' expression
-  IDecIntArithOp   :: ArithOp             -> Invariant Decimal -> Invariant Integer -> Invariant Decimal
-  -- | An 'ArithOp' expression over an 'Integer' and a 'Decimal' expression
-  IIntDecArithOp   :: ArithOp             -> Invariant Integer -> Invariant Decimal -> Invariant Decimal
-  -- | Integer modulus
-  IModOp           :: Invariant Integer   -> Invariant Integer ->                      Invariant Integer
-  -- | Rounds a 'Decimal' expression to an 'Integer' expression per the
-  -- 'RoundingLikeOp' strategy
-  IRoundingLikeOp1 :: RoundingLikeOp      -> Invariant Decimal ->                      Invariant Integer
-  -- | Rounds a 'Decimal' expression to a 'Decimal' expression with the
-  -- specified 'Integer' level of precision and per the 'RoundingLikeOp'
-  -- strategy
-  IRoundingLikeOp2 :: RoundingLikeOp      -> Invariant Decimal -> Invariant Integer -> Invariant Decimal
+data EInvariant where
```

Comment:
all three terms now have an existential representation. also, for free we get that invariants can talk about objects now.

Reply:
this is great